Firestore/core/include/firebase/firestore/document_reference.h

```diff
@@ -0,0 +1,362 @@
+// removed once this note has migrated to the API review doc.
+
+// TODO(rsgowman): replace these forward decl's with appropriate includes (once
+// they exist)
+namespace firebase {
+class App;
+template <typename T>
+class Future;
+}  // namespace firebase
+
+namespace firebase {
+namespace firestore {
+
+// TODO(rsgowman): replace these forward decl's with appropriate includes (once
+// they exist)
+class FieldValue;
+class DocumentSnapshot;
+class Firestore;
+class Error;
+template <typename T>
+class EventListener;
+class ListenerRegistration;
+class CollectionReference;
+class DocumentListenOptions;
+// TODO(rsgowman): not quite a forward decl, but required to make the default
+// parameter to Set() "compile".
+class SetOptions {
+ public:
+  SetOptions();
+};
+
+// TODO(rsgowman): move this into the FieldValue header
+#ifdef STLPORT
+using MapFieldValue = std::tr1::unordered_map<std::string, FieldValue>;
+#else
+using MapFieldValue = std::unordered_map<std::string, FieldValue>;
+#endif
+
+/**
+ * A DocumentReference refers to a document location in a Firestore database and
+ * can be used to write, read, or listen to the location. There may or may not
+ * exist a document at the referenced location. A DocumentReference can also be
+ * used to create a CollectionReference to a subcollection.
+ *
+ * Create a DocumentReference via Firebase::Document(const string& path).
+ *
+ * Subclassing Note: Firestore classes are not meant to be subclassed except for
+ * use in test mocks. Subclassing is not supported in production code and new
+ * SDK releases may break code that does so.
+ */
+class DocumentReference {
+ public:
+  /**
+   * @brief Default constructor. This creates an invalid DocumentReference.
+   * Attempting to perform any operations on this reference will fail (and cause
+   * a crash) unless a / valid DocumentReference has been assigned to it.
+   */
+  DocumentReference();
+
+  /**
+   * @brief Copy constructor. It's totally okay (and efficient) to copy
+   * DocumentReference instances, as they simply point to the same location in
+   * the database.
+   *
+   * @param[in] reference DocumentReference to copy from.
+   */
+  DocumentReference(const DocumentReference& reference);
+
+  /**
+   * @brief Move constructor. Moving is an efficient operation for
+   * DocumentReference instances.
+   *
+   * @param[in] reference DocumentReference to move data from.
+   */
+  DocumentReference(DocumentReference&& reference);
+
+  virtual ~DocumentReference();
+
+  /**
+   * @brief Copy assignment operator. It's totally okay (and efficient) to copy
+   * DocumentReference instances, as they simply point to the same location in
+   * the database.
+   *
+   * @param[in] reference DocumentReference to copy from.
+   *
+   * @returns Reference to the destination DocumentReference.
+   */
+  virtual DocumentReference& operator=(const DocumentReference& reference);
```

Comment:
Virtual assignment operator is a little unusual. How is it supposed to be used?

Reply:
Oops. I blindly added 'virtual' everywhere to allow mocking. Fixed. (Also removed it from the move assignment operator immediately below.)